File: research/ntrip_ingestion/src/nlgcp_ntrip_ingest/subjects.py
"""Downstream delivery subjects for live frames.

Live ingestion publishes — when a transport is attached — under::

    correction.live.<STATION_ID>

mirroring the Phase 9 ``correction.replay.<STATION_ID>`` family so a
single consumer contract covers both origins.  No transport is opened
by this module; names only.
"""

from __future__ import annotations

from nlgcp_ntrip_ingest.models import LiveFrame

SUBJECT_FAMILY = "correction"
LIVE_SUBJECT_PREFIX = "correction.live"


def live_subject(station_id: str) -> str:
    """Subject for live frames of one station (no transport attached)."""
    station = station_id.strip().upper()
    if not station:
        raise ValueError("station_id must be non-empty")
    if any(c not in "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_" for c in station):
        raise ValueError(f"unsafe station_id for subject: {station_id}")
    return f"{LIVE_SUBJECT_PREFIX}.{station}"


def ordering_key(frame: LiveFrame) -> tuple[str, str, int]:
    """Stable ordering key: (connection_id, capture order, sequence)."""
    return (frame.connection_id, frame.provenance.get("capture_id", ""), frame.sequence)


def envelope(frame: LiveFrame) -> dict[str, object]:
    """Delivery envelope: ordering key + provenance, no file internals."""
    return {
        "subject": live_subject(frame.station_id),
        "ordering_key": list(ordering_key(frame)),
        "frame": frame.as_dict(),
    }

File: services/gnss-ingestor/internal/ntrip/subjects.go
package ntrip

import (
	"fmt"
	"strings"
)

// Live delivery subjects mirror the reserved correction.* family:
// live ingestion publishes under correction.live.<STATION> while the
// Phase 9 replay path uses correction.replay.<STATION>, so one consumer
// contract covers both origins. Names only; no transport is opened here.
const (
	SubjectFamily      = "correction"
	LiveSubjectPrefix  = "correction.live"
	ReplaySubjectPrime = "correction.replay"
)

// LiveSubject returns the delivery subject for one station.
func LiveSubject(stationID string) (string, error) {
	station := strings.ToUpper(strings.TrimSpace(stationID))
	if station == "" {
		return "", &ConfigError{Field: "station_id"}
	}
	for _, r := range station {
		if !(r >= 'A' && r <= 'Z' || r >= '0' && r <= '9' || r == '-' || r == '_') {
			return "", fmt.Errorf("unsafe station_id for subject: %q", stationID)
		}
	}
	return LiveSubjectPrefix + "." + station, nil
}
